pvlib/tests/test_singlediode.py

```diff
@@ -69,6 +71,123 @@ def test_newton_fs_495(method, cec_module_fs_495):
+    ``Vth``. The columns ``Irradiance``, ``Sweep Direction`` are None or empty
+    strings.
+
+    Parameters
+    ----------
+    file_csv: str
+        Path to a CSV file of iv curve parameter sets.
+
+    file_json: str
+        Path to a JSON file of precise iv curves.
+
+    Returns
+    -------
+        A data frame.
+    """
+    params = pd.read_csv(file_csv)
+    curves_metadata = pd.read_json(file_json)
+    curves = pd.DataFrame(curves_metadata['IV Curves'].values.tolist())
+    curves['cells_in_series'] = curves_metadata['cells_in_series']
+    joined = params.merge(curves, on='Index', how='inner',
+                          suffixes=(None, '_drop'), validate='one_to_one')
+    joined = joined[(c for c in joined.columns if not c.endswith('_drop'))]
+
+    # parse strings to np.float64
+    is_array = ['Currents', 'Voltages', 'diode_voltage']
+    joined[is_array] = joined[is_array].applymap(
+        lambda a: np.asarray(a, dtype=np.float64)
+    )
+    is_number = ['v_oc', 'i_sc', 'v_mp', 'i_mp', 'p_mp', 'i_x', 'i_xx',
+                 'Temperature']
+    joined[is_number] = joined[is_number].applymap(np.float64)
+
+    joined['Boltzman'] = 1.380649e-23
+    joined['Electron Charge'] = 1.60217663e-19
+    joined['Vth'] = (
+        joined['Boltzman'] * joined['Temperature']
+        / joined['Electron Charge']
+    )
+
+    return joined
+
+
+@pytest.fixture(scope='function', params=[
+    {
+        'csv': f'{DATA_DIR}/precise_iv_curves_parameter_sets1.csv',
+        'json': f'{DATA_DIR}/precise_iv_curves1.json'
+    },
+    {
+        'csv': f'{DATA_DIR}/precise_iv_curves_parameter_sets2.csv',
+        'json': f'{DATA_DIR}/precise_iv_curves2.json'
+    }
+], ids=[1, 2])
+def precise_iv_curves(request):
+    file_csv, file_json = request.param['csv'], request.param['json']
+    pc = build_precise_iv_curve_dataframe(file_csv, file_json)
+    params = ['photocurrent', 'saturation_current', 'resistance_series',
+              'resistance_shunt']
+    singlediode_params = pc.loc[:, params]
+    singlediode_params['nNsVth'] = pc['n'] * pc['cells_in_series'] * pc['Vth']
+    return singlediode_params, pc
+
+
+@pytest.mark.parametrize('method', ['lambertw', 'brentq', 'newton'])
+def test_singlediode_precision(method, precise_iv_curves):
+    """
+    Tests the accuracy of singlediode. ivcurve_pnts is not tested.
+    """
+    x, pc = precise_iv_curves
+    outs = pvsystem.singlediode(method=method, **x)
+
+    assert np.allclose(pc['i_sc'], outs['i_sc'], atol=1e-10, rtol=0)
+    assert np.allclose(pc['v_oc'], outs['v_oc'], atol=1e-10, rtol=0)
+    assert np.allclose(pc['i_mp'], outs['i_mp'], atol=5e-8, rtol=0)
+    assert np.allclose(pc['v_mp'], outs['v_mp'], atol=8e-7, rtol=0)
+    assert np.allclose(pc['p_mp'], outs['p_mp'], atol=1e-10, rtol=0)
+    assert np.allclose(pc['i_x'], outs['i_x'], atol=1e-10, rtol=0)
+    assert np.allclose(pc['i_xx'], outs['i_xx'], atol=8e-8, rtol=0)
```

Comment:
I'm super curious how these tolerances might be tightened for 'lambertw', 'brentq', vs. 'newton'.

Reply:
And I'm curious why there are test failures only on ubuntu and python <=3.9 for the 'lambertw' case. I suspect something buried fairly deep behind `scipy.special.lambertw` which we won't solve.

Reply:
For [this failed test](https://github.com/pvlib/pvlib-python/actions/runs/3879773557/jobs/6617268775), the 30th `i_xx` value differs by `1e-6`. All of the tests related to this pull request pass with `atol=1e-6` for the `i_xx` value.

Reply:
@reepoi I think there's a mistake in the precise curve library somewhere.